### DISCUSSION TITLE
Predefined queries

### DISCUSSION THREAD

**Discussion by drnextgis** [stac-utils/titiler-pgstac#33] [Q&A] [answered]
Hi! Could I ask for guidance regarding the following issue.

By default titiler-pgstac accepts an arbitrary search query for `/mosaic/register` endpoint but I want to do it more restrictive. Let's assume that for every request based on the information provided inside authorization token we can build our internal [SearchQuery](https://github.com/stac-utils/titiler-pgstac/blob/13ff815bd1b793a5ae22bc2b42770002030ead94/titiler/pgstac/models.py#L57) object and combine it with the one which is coming from a request using *AND* clause. Do you think it can be implemented in a general way that other users of titiler-pgstac can benefit from or it is very specific task which requires creation of a tailored application. Thanks!

**Comment by vincentsarago** [comment, marked as answer]
@drnextgis thanks for opening the discussion. 

I don't think there is an easy solution other than for you to re-write https://github.com/stac-utils/titiler-pgstac/blob/master/titiler/pgstac/factory.py#L338-L366

**Reply by drnextgis** [reply]
I was thinking about something like this:

```diff
diff --git a/titiler/pgstac/factory.py b/titiler/pgstac/factory.py
index ac39511..0c54855 100644
--- a/titiler/pgstac/factory.py
+++ b/titiler/pgstac/factory.py
@@ -70,6 +70,8 @@ class MosaicTilerFactory(BaseTilerFactory):
 
     backend_options: Dict = field(default_factory=dict)
 
+    query_factory: Callable[..., SearchQuery] = None
+
     def register_routes(self) -> None:
         """This Method register routes to the router."""
         self._search_routes()
@@ -341,13 +343,18 @@ class MosaicTilerFactory(BaseTilerFactory):
         )
         def register_search(request: Request, body: SearchQuery):
             """Register a Search query."""
-            search = body.json(
+            search = body.dict(
                 exclude_none=True,
                 exclude={"metadata"},
                 by_alias=True,
             )
             metadata = body.metadata or {}
 
+            if self.query_factory is not None:
+                search = self.query_factory(request, search)
+
+            search = json.dumps(search)
+
             with request.app.state.dbpool.connection() as conn:
                 with conn.cursor() as cursor:
                     cursor.execute(
```

Then user can create custom `query_factory` function and pass it to `MosaicTilerFactory` constructor:

```diff
diff --git a/titiler/pgstac/main.py b/titiler/pgstac/main.py
index c38f143..6b2aba4 100644
--- a/titiler/pgstac/main.py
+++ b/titiler/pgstac/main.py
@@ -67,7 +67,19 @@ if settings.debug:
 else:
     optional_headers = []
 
-mosaic = MosaicTilerFactory(optional_headers=optional_headers)
+
+def query_factory(request, search):
+    if "filter" in search:
+        search["filter"] = {
+            "op": "and",
+            "args": [
+                {"op": ">", "args": [{"property": "eo:cloud_cover"}, 70]},
+                search["filter"]
+            ]
+        }
+    return search
+
+mosaic = MosaicTilerFactory(optional_headers=optional_headers, query_factory=query_factory)
 app.include_router(mosaic.router)
 
 tms = TMSFactory(supported_tms=TileMatrixSetName, tms_dependency=TMSParams)
```

**Reply by vincentsarago** [reply]
looks good to me, I think it will be fine if the factory was `search_dependency` was set to a default callable like https://github.com/developmentseed/titiler/blob/master/src/titiler/core/titiler/core/factory.py#L123

**Reply by drnextgis** [reply]
Thank you @vincentsarago! I've made a PR https://github.com/stac-utils/titiler-pgstac/pull/41.
